internal/mesh/internal/types/intermediate/types.go

```diff
@@ -0,0 +1,61 @@
+package intermediate
+
+import (
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// todo should it be destination?
+// the problem is that it's compiled from different source objects
+type CombinedDestinationRef struct {
+	// ServiceRef is the reference to the destination service for this upstream
+	ServiceRef *pbresource.Reference
+
+	Port string
+
+	// sourceProxies are the IDs of source proxy state template resources.
+	SourceProxies map[string]*pbresource.ID
+
+	// explicitUpstreamID is the id of an explicit upstreams resource. For implicit upstreams,
+	// this should be nil.
+	ExplicitDestinationsID *pbresource.ID
+}
+
+type ServiceEndpoints struct {
```

Comment:
This is very similar to my `wrappedtypes` package: https://github.com/hashicorp/consul-enterprise/blob/dc82a8cc96261c90dc1e7cd573df77b3d917ca64/internal/mesh/internal/wrappedtypes/wrappedtypes.go